packages/datadog-plugin-aws-sdk/src/index.js

```diff
@@ -3,67 +3,100 @@
 
-        awsHelpers.addResponseTags(span, response, serviceIdentifier, config, tracer)
-        awsHelpers.finish(config, span, response, response.error)
-      })
+      awsHelpers.requestInject(span, request, serviceIdentifier, this.tracer)
 
-      analyticsSampler.sample(span, config.measured)
+      const store = storage.getStore()
 
-      awsHelpers.requestInject(span, this, serviceIdentifier, tracer)
+      this.enter(span, store)
+    })
 
-      const request = this
+    this.addSub('apm:aws:request:complete', ({ response, serviceIdentifier }) => {
+      const store = storage.getStore()
+      if (!store) return
+      const { span } = store
+      if (!span) return
+      awsHelpers.addResponseTags(span, response, serviceIdentifier)
+      awsHelpers.finish(this.config, span, response, response.error)
+    })
 
-      return tracer.scope().activate(span, () => {
-        if (typeof cb === 'function') {
-          arguments[0] = awsHelpers.wrapCb(cb, serviceIdentifier, tags, request, tracer, childOf)
+    this.addSub('apm:aws:response', obj => {
+      const { request, response, serviceName } = obj
+      const store = storage.getStore()
+      const plugin = this
+      const maybeChildOf = awsHelpers.responseExtract(serviceName, request, response, this.tracer)
+      if (maybeChildOf) {
+        const options = {
+          childOf: maybeChildOf,
+          tags: Object.assign({}, requestTags.get(request) || {}, { [Tags.SPAN_KIND]: 'server' })
         }
-        return send.apply(this, arguments)
-      })
-    }
+        obj.ar = {
+          real: new AsyncResource('apm:aws:response'),
+          runInAsyncScope (fn) {
+            return this.real.runInAsyncScope(() => {
+              const span = plugin.tracer.startSpan('aws.response', options)
+              plugin.enter(span, store)
+              try {
+                let result = fn()
+                if (result && result.then) {
+                  result = result.then(x => {
+                    span.finish()
+                    return x
+                  }, e => {
+                    span.setTag(e)
+                    throw e
+                  })
+                } else {
+                  span.finish()
+                }
+                return result
+              } catch (e) {
+                span.setTag(e)
+                span.finish()
+              }
+            })
+          }
+        }
+      }
+    })
   }
-}
-
-function createWrapSetPromisesDependency (tracer, config, instrumenter, AWS) {
-  return function wrapSetPromisesDependency (setPromisesDependency) {
-    return function setPromisesDependencyWithTrace (dep) {
-      const result = setPromisesDependency.apply(this, arguments)
 
-      instrumenter.wrap(AWS.Request.prototype, 'promise', createWrapRequest(tracer, config))
-
-      return result
-    }
+  configure (config) {
+    Plugin.prototype.configure.call(this, config)
+    this.config = normalizeConfig(config)
```

Comment:
This ends up assigning the config twice. The result of `normalizeConfig` could be passed to super instead.